File: examples/hako_drone/src/drone_cmd.c
#include "hako_asset.h"
#include "hako_conductor.h"
#include <stdio.h>
#include <stdlib.h>
#include "hako_msgs/pdu_ctype_HakoDroneCmdTakeoff.h"
#include "hako_msgs/pdu_ctype_HakoDroneCmdMove.h"
#include "hako_msgs/pdu_ctype_HakoDroneCmdLand.h"
#define HAKO_AVATOR_CHANNLE_ID_MOTOR        0
#define HAKO_AVATOR_CHANNLE_ID_POS          1
#define HAKO_AVATOR_CHANNLE_ID_COLLISION    2
#define HAKO_AVATOR_CHANNLE_ID_MANUAL       3
#define HAKO_AVATOR_CHANNEL_ID_DISTURB      4
#define HAKO_AVATOR_CHANNEL_ID_CMD_TAKEOFF  5
#define HAKO_AVATOR_CHANNEL_ID_CMD_MOVE     6
#define HAKO_AVATOR_CHANNEL_ID_CMD_LAND     7


#ifdef _WIN32
static inline void usleep(long microseconds) {
    Sleep(microseconds / 1000);
}
#else
#include <unistd.h>
#endif

static void do_cmd_takeoff(void)
{
    printf("INFO: start takeoff\n");
    int result = 0;
    Hako_HakoDroneCmdTakeoff cmd_takeoff;
    cmd_takeoff.header.request = 1;
    cmd_takeoff.header.result = 0;
    cmd_takeoff.header.result_code = 0;
    cmd_takeoff.height = 5;
    cmd_takeoff.speed = 5;
    hako_asset_pdu_write("DroneFlightController", HAKO_AVATOR_CHANNEL_ID_CMD_TAKEOFF, (const char*)&cmd_takeoff, sizeof(cmd_takeoff));
    while (result == 0) {
        if (hako_asset_pdu_read("DroneFlightController", HAKO_AVATOR_CHANNEL_ID_CMD_TAKEOFF, (char*)&cmd_takeoff, sizeof(cmd_takeoff)) == 0) {
            if (cmd_takeoff.header.result) {
                cmd_takeoff.header.result = 0;
                hako_asset_pdu_write("DroneFlightController", HAKO_AVATOR_CHANNEL_ID_CMD_TAKEOFF, (const char*)&cmd_takeoff, sizeof(cmd_takeoff));
                printf("DONE\n");
                break;
            }
        }
        usleep(1000*100);
    }
}
static void do_cmd_move(double x, double y)
{
    printf("INFO: start move\n");
    int result = 0;
    Hako_HakoDroneCmdMove cmd;
    cmd.header.request = 1;
    cmd.header.result = 0;
    cmd.header.result_code = 0;
    cmd.x = x;
    cmd.y = y;
    cmd.speed = 5;
    hako_asset_pdu_write("DroneFlightController", HAKO_AVATOR_CHANNEL_ID_CMD_MOVE, (const char*)&cmd, sizeof(cmd));
    while (result == 0) {
        if (hako_asset_pdu_read("DroneFlightController", HAKO_AVATOR_CHANNEL_ID_CMD_MOVE, (char*)&cmd, sizeof(cmd)) == 0) {
            if (cmd.header.result) {
                cmd.header.result = 0;
                hako_asset_pdu_write("DroneFlightController", HAKO_AVATOR_CHANNEL_ID_CMD_MOVE, (const char*)&cmd, sizeof(cmd));
                printf("DONE\n");
                break;
            }
        }
        usleep(1000*100);
    }
}
static void do_cmd_land(void)
{
    printf("INFO: start land\n");
    int result = 0;
    Hako_HakoDroneCmdLand cmd;
    cmd.header.request = 1;
    cmd.header.result = 0;
    cmd.header.result_code = 0;
    cmd.height = 0;
    cmd.speed = 5;
    hako_asset_pdu_write("DroneFlightController", HAKO_AVATOR_CHANNEL_ID_CMD_LAND, (const char*)&cmd, sizeof(cmd));
    while (result == 0) {
        if (hako_asset_pdu_read("DroneFlightController", HAKO_AVATOR_CHANNEL_ID_CMD_LAND, (char*)&cmd, sizeof(cmd)) == 0) {
            if (cmd.header.result) {
                cmd.header.result = 0;
                hako_asset_pdu_write("DroneFlightController", HAKO_AVATOR_CHANNEL_ID_CMD_LAND, (const char*)&cmd, sizeof(cmd));
                printf("DONE\n");
                break;
            }
        }
        usleep(1000*100);
    }
}
int main(int argc, const char* argv[])
{
    int ret = hako_initialize_for_external();
    if (ret != 0) {
        printf("ERORR: hako_initialize_for_external() returns %d.", ret);
        return 1;
    }
    do_cmd_takeoff();
    do_cmd_move(0, 5);
    //do_cmd_move(-5, -5);
    //do_cmd_move(0, 0);
    do_cmd_land();
    return 0;
}

File: examples/hako_drone/drone_cmd.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

import hakopy
import hako_pdu
import pdu_info
import sys
import time

pdu_manager = None
def do_takeoff():
    global pdu_manager
    command = pdu_manager.get_pdu('DroneFlightController', pdu_info.HAKO_AVATOR_CHANNEL_ID_CMD_TAKEOFF)
    pdu_cmd = command.get()
    pdu_cmd['header']['request'] = 1
    pdu_cmd['header']['result'] = 0
    pdu_cmd['header']['result_code'] = 0
    pdu_cmd['height'] = 5
    pdu_cmd['speed'] = 5
    ret = command.write()
    if ret == False:
        print('"ERROR: hako_asset_pdu_write')
        return 0
    
    while True:
        pdu = command.read()
        if pdu == None:
            print('ERROR: hako_asset_pdu_read')
            return 0
        if pdu['header']['result'] == 1:
            pdu['header']['result'] = 0
            command.write()
            print('DONE')
            break
    print("INFO: do_takeoff exit")
    return 0


def do_move(x, y):
    global pdu_manager
    command = pdu_manager.get_pdu('DroneFlightController', pdu_info.HAKO_AVATOR_CHANNEL_ID_CMD_MOVE)
    pdu_cmd = command.get()
    pdu_cmd['header']['request'] = 1
    pdu_cmd['header']['result'] = 0
    pdu_cmd['header']['result_code'] = 0
    pdu_cmd['x'] = x
    pdu_cmd['y'] = y
    pdu_cmd['speed'] = 5
    ret = command.write()
    if ret == False:
        print('"ERROR: hako_asset_pdu_write')
        return 0
    
    while True:
        pdu = command.read()
        if pdu == None:
            print('ERROR: hako_asset_pdu_read')
            return 0
        if pdu['header']['result'] == 1:
            pdu['header']['result'] = 0
            command.write()
            print('DONE')
            break
    print("INFO: do_move exit")
    return 0



def do_land():
    global pdu_manager
    command = pdu_manager.get_pdu('DroneFlightController', pdu_info.HAKO_AVATOR_CHANNEL_ID_CMD_LAND)
    pdu_cmd = command.get()
    pdu_cmd['header']['request'] = 1
    pdu_cmd['header']['result'] = 0
    pdu_cmd['header']['result_code'] = 0
    pdu_cmd['height'] = 0
    pdu_cmd['speed'] = 5
    ret = command.write()
    if ret == False:
        print('"ERROR: hako_asset_pdu_write')
        return 0
    
    while True:
        pdu = command.read()
        if pdu == None:
            print('ERROR: hako_asset_pdu_read')
            return 0
        if pdu['header']['result'] == 1:
            pdu['header']['result'] = 0
            command.write()
            print('DONE')
            break
    print("INFO: do_land exit")
    return 0


def main():
    global pdu_manager
    if len(sys.argv) != 2:
        print(f"Usage: {sys.argv[0]} <config_path>")
        return 1

    config_path = sys.argv[1]

    pdu_manager = hako_pdu.HakoPduManager('/usr/local/lib/hakoniwa/hako_binary/offset', config_path)
    ret = hakopy.init_for_external()
    if ret == False:
        print(f"ERROR: init_for_external() returns {ret}.")
        return 1
    do_takeoff()
    do_move(0, 5)
    do_land()
    return 0

if __name__ == "__main__":
    sys.exit(main())
